File: ios/TNGuardTricare/Stores/ProgressStore.swift
import Foundation

/// Tracks which checklist steps and forms the user has marked complete. Non-sensitive —
/// just step IDs — so it's stored as a plain JSON file in the app's Documents directory.
/// Sensitive form field values (SSN, bank info) live in KeychainStore instead.
@MainActor
final class ProgressStore: ObservableObject {
    @Published private(set) var completedTRSSteps: Set<String>
    @Published private(set) var completedReimbursementSteps: Set<String>
    @Published private(set) var completedForms: Set<String>

    private struct State: Codable {
        var completedTRSSteps: Set<String> = []
        var completedReimbursementSteps: Set<String> = []
        var completedForms: Set<String> = []
    }

    private var fileURL: URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("progress.json")
    }

    init() {
        let loaded = Self.load(from: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("progress.json"))
        completedTRSSteps = loaded.completedTRSSteps
        completedReimbursementSteps = loaded.completedReimbursementSteps
        completedForms = loaded.completedForms
    }

    func isStepComplete(_ id: String, track: Track) -> Bool {
        switch track {
        case .trs: return completedTRSSteps.contains(id)
        case .reimbursement: return completedReimbursementSteps.contains(id)
        }
    }

    func toggleStep(_ id: String, track: Track) {
        switch track {
        case .trs:
            if completedTRSSteps.contains(id) { completedTRSSteps.remove(id) } else { completedTRSSteps.insert(id) }
        case .reimbursement:
            if completedReimbursementSteps.contains(id) { completedReimbursementSteps.remove(id) } else { completedReimbursementSteps.insert(id) }
        }
        persist()
    }

    func setStepComplete(_ id: String, track: Track, complete: Bool) {
        switch track {
        case .trs:
            if complete { completedTRSSteps.insert(id) } else { completedTRSSteps.remove(id) }
        case .reimbursement:
            if complete { completedReimbursementSteps.insert(id) } else { completedReimbursementSteps.remove(id) }
        }
        persist()
    }

    func isFormComplete(_ id: String) -> Bool { completedForms.contains(id) }

    func setFormComplete(_ id: String, complete: Bool) {
        if complete { completedForms.insert(id) } else { completedForms.remove(id) }
        persist()
    }

    func progressFraction(track: Track, totalSteps: Int) -> Double {
        guard totalSteps > 0 else { return 0 }
        let completedCount: Int
        switch track {
        case .trs: completedCount = completedTRSSteps.count
        case .reimbursement: completedCount = completedReimbursementSteps.count
        }
        return min(1.0, Double(completedCount) / Double(totalSteps))
    }

    func resetAll() {
        completedTRSSteps = []
        completedReimbursementSteps = []
        completedForms = []
        persist()
    }

    enum Track: Hashable {
        case trs
        case reimbursement
    }

    private func persist() {
        let state = State(
            completedTRSSteps: completedTRSSteps,
            completedReimbursementSteps: completedReimbursementSteps,
            completedForms: completedForms
        )
        guard let data = try? JSONEncoder().encode(state) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }

    private static func load(from url: URL) -> State {
        guard let data = try? Data(contentsOf: url),
              let state = try? JSONDecoder().decode(State.self, from: data) else {
            return State()
        }
        return state
    }
}

File: android/app/src/main/java/org/tnguardtricare/app/data/ProgressStore.kt
package org.tnguardtricare.app.data

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringSetPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

private val Context.progressDataStore by preferencesDataStore(name = "progress")

enum class Track { TRS, REIMBURSEMENT }

/**
 * Tracks which checklist steps and forms the user has marked complete. Non-sensitive — just
 * step IDs — so it's stored in plain DataStore preferences. Sensitive form field values (SSN,
 * bank info) live in SecureFieldStore instead. Mirrors iOS's ProgressStore.swift.
 */
class ProgressStore(private val context: Context) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    private val keyTrs = stringSetPreferencesKey("completed_trs_steps")
    private val keyReimbursement = stringSetPreferencesKey("completed_reimbursement_steps")
    private val keyForms = stringSetPreferencesKey("completed_forms")

    private val _completedTrsSteps = MutableStateFlow<Set<String>>(emptySet())
    val completedTrsSteps: StateFlow<Set<String>> = _completedTrsSteps.asStateFlow()

    private val _completedReimbursementSteps = MutableStateFlow<Set<String>>(emptySet())
    val completedReimbursementSteps: StateFlow<Set<String>> = _completedReimbursementSteps.asStateFlow()

    private val _completedForms = MutableStateFlow<Set<String>>(emptySet())
    val completedForms: StateFlow<Set<String>> = _completedForms.asStateFlow()

    init {
        scope.launch {
            context.progressDataStore.data.collect { prefs ->
                _completedTrsSteps.value = prefs[keyTrs] ?: emptySet()
                _completedReimbursementSteps.value = prefs[keyReimbursement] ?: emptySet()
                _completedForms.value = prefs[keyForms] ?: emptySet()
            }
        }
    }

    fun isStepComplete(id: String, track: Track): Boolean {
        val set = if (track == Track.TRS) _completedTrsSteps.value else _completedReimbursementSteps.value
        return set.contains(id)
    }

    fun toggleStep(id: String, track: Track) {
        val current = if (track == Track.TRS) _completedTrsSteps.value else _completedReimbursementSteps.value
        setStepComplete(id, track, !current.contains(id))
    }

    fun setStepComplete(id: String, track: Track, complete: Boolean) {
        scope.launch {
            context.progressDataStore.edit { prefs ->
                val key = if (track == Track.TRS) keyTrs else keyReimbursement
                val current = prefs[key] ?: emptySet()
                prefs[key] = if (complete) current + id else current - id
            }
        }
    }

    fun isFormComplete(id: String): Boolean = _completedForms.value.contains(id)

    fun setFormComplete(id: String, complete: Boolean) {
        scope.launch {
            context.progressDataStore.edit { prefs ->
                val current = prefs[keyForms] ?: emptySet()
                prefs[keyForms] = if (complete) current + id else current - id
            }
        }
    }

    fun progressFraction(track: Track, totalSteps: Int): Float {
        if (totalSteps <= 0) return 0f
        val completedCount = if (track == Track.TRS) _completedTrsSteps.value.size else _completedReimbursementSteps.value.size
        return (completedCount.toFloat() / totalSteps.toFloat()).coerceAtMost(1f)
    }

    fun resetAll() {
        scope.launch {
            context.progressDataStore.edit { it.clear() }
        }
    }
}
